packages/mdl-ripple/mixin.js

```diff
@@ -0,0 +1,313 @@
+        let radiusDuration = 1000 * Math.sqrt(remainingRadius / 4424);
+        let originDuration = radiusDuration;
+
+        let opacityDuration = 1000 * circleOpacity / 3;
+        this.renderer_.setStyles(fgCircle, {
+          transition: `opacity ${opacityDuration}ms linear, transform ${radiusDuration}ms ${LOG_DECELERATE}`,
+          transform: `scale(${1 + TRANSITION_DELTA})`,
+          opacity: '0'
+        });
+
+        this.renderer_.setStyles(foreground, {
+          transition: `transform ${originDuration}ms ${LOG_DECELERATE}`,
+          transform: `translate3d(0,0,0)`
+        });
+
+        // Since we know the duration already, no need to listen for transition end.
+        setTimeout(() => {
+          this.resetStyles_(foreground, fgCircle);
+        }, opacityDuration);
+      }
+
+      this.fgIndex_ = (this.fgIndex_ + 1) % MAX_RIPPLES;
+    },
+
+
+    /**
+     * @param {!Array<ElementRef>} eles
+     */
+    resetStyles_(...eles) {
+      eles.forEach((each) => {
+        this.renderer_.setStyles(each, {
+          transition: '',
+          transform: '',
+          opacity: ''
+        });
+      });
+    },
+
+
+    cancelAnimations() {
+      let background = this.renderer_.get(Identifier.BACKGROUND);
+      this.renderer_.removeClass(background, Class.BACKGROUND_ACTIVE);
+
+      let foreground = this.renderer_.get(Identifier.FOREGROUND, this.fgIndex_);
+      let fgCircle = this.renderer_.get(Identifier.FOREGROUND_CIRCLE, this.fgIndex_);
+      this.resetStyles_(foreground, fgCircle);
+    }
+  });
+
+
+  Object.defineProperty(this, 'maxRadius', {
+    get: function() {
+      if (!this.maxRadius_) {
+        let maxDim = Math.max(this.boundingRect_.height, this.boundingRect_.width);
+        // Sqrt(2) * square length == diameter
+        return Math.sqrt(2) * maxDim / 2;
+      } else {
+        return this.maxRadius_;
+      }
+    },
+
+    set: function(value) {
+      this.maxRadius_ = value;
+    }
+  });
+}
+
+
+/**
+ * @param {!Event} ev
+ * @param {!Element} container
+ * @return {{top: number, left:number}}
+ */
+export function getNormalizedEventCoords(ev, container) {
+  let rect = container.getBoundingClientRect();
+  let documentLeft =  window.pageXOffset + rect.left;
+  let documentTop =  window.pageYOffset + rect.top;
+
+  var normalizedLeft;
+  var normalizedTop;
+  // Determine touch point relative to the ripple container.
+  if (ev.type == 'touchstart') {
+    normalizedLeft = ev.touches[0].pageX - documentLeft;
+    normalizedTop = ev.touches[0].pageY - documentTop;
+  } else {
+    normalizedLeft = ev.pageX - documentLeft;
+    normalizedTop = ev.pageY - documentTop;
+  }
+
+  return { left: normalizedLeft, top: normalizedTop };
+};
+
+
+/**
+ * Create a cloned object with functions bound to the given this parameter.
+ * @param {!Object} obj
+ * @param {!Object} self
+ * @return {!Object}
+ */
+function bindAll(obj, self) {
+  let boundObj = {};
+  for (let p in obj) {
```

Comment:
maybe be better to use `Object.keys()` here. Or if not use a `hasOwnProperty()` check